# Наивный байесовский классификатор

In [1]:
from sklearn import datasets
from sklearn.cross_validation import cross_val_score

C:\anaconda\lib\site-packages\sklearn\cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [2]:
digit = datasets.load_digits()
breast_cancer = datasets.load_breast_cancer()

In [3]:
digit.data[:3]

array([[  0.,   0.,   5.,  13.,   9.,   1.,   0.,   0.,   0.,   0.,  13.,
         15.,  10.,  15.,   5.,   0.,   0.,   3.,  15.,   2.,   0.,  11.,
          8.,   0.,   0.,   4.,  12.,   0.,   0.,   8.,   8.,   0.,   0.,
          5.,   8.,   0.,   0.,   9.,   8.,   0.,   0.,   4.,  11.,   0.,
          1.,  12.,   7.,   0.,   0.,   2.,  14.,   5.,  10.,  12.,   0.,
          0.,   0.,   0.,   6.,  13.,  10.,   0.,   0.,   0.],
       [  0.,   0.,   0.,  12.,  13.,   5.,   0.,   0.,   0.,   0.,   0.,
         11.,  16.,   9.,   0.,   0.,   0.,   0.,   3.,  15.,  16.,   6.,
          0.,   0.,   0.,   7.,  15.,  16.,  16.,   2.,   0.,   0.,   0.,
          0.,   1.,  16.,  16.,   3.,   0.,   0.,   0.,   0.,   1.,  16.,
         16.,   6.,   0.,   0.,   0.,   0.,   1.,  16.,  16.,   6.,   0.,
          0.,   0.,   0.,   0.,  11.,  16.,  10.,   0.,   0.],
       [  0.,   0.,   0.,   4.,  15.,  12.,   0.,   0.,   0.,   0.,   3.,
         16.,  15.,  14.,   0.,   0.,   0.,   0.,   8.,  13.

In [4]:
breast_cancer.data[:3]

array([[  1.79900000e+01,   1.03800000e+01,   1.22800000e+02,
          1.00100000e+03,   1.18400000e-01,   2.77600000e-01,
          3.00100000e-01,   1.47100000e-01,   2.41900000e-01,
          7.87100000e-02,   1.09500000e+00,   9.05300000e-01,
          8.58900000e+00,   1.53400000e+02,   6.39900000e-03,
          4.90400000e-02,   5.37300000e-02,   1.58700000e-02,
          3.00300000e-02,   6.19300000e-03,   2.53800000e+01,
          1.73300000e+01,   1.84600000e+02,   2.01900000e+03,
          1.62200000e-01,   6.65600000e-01,   7.11900000e-01,
          2.65400000e-01,   4.60100000e-01,   1.18900000e-01],
       [  2.05700000e+01,   1.77700000e+01,   1.32900000e+02,
          1.32600000e+03,   8.47400000e-02,   7.86400000e-02,
          8.69000000e-02,   7.01700000e-02,   1.81200000e-01,
          5.66700000e-02,   5.43500000e-01,   7.33900000e-01,
          3.39800000e+00,   7.40800000e+01,   5.22500000e-03,
          1.30800000e-02,   1.86000000e-02,   1.34000000e-02,
       

In [5]:
from sklearn.naive_bayes import BernoulliNB, GaussianNB, MultinomialNB
bernulli = BernoulliNB()
gaussian = GaussianNB()
multinomial = MultinomialNB()

## Бернулли

In [6]:
cross_val_score(estimator = bernulli,X = digit.data,y = digit.target).mean()

0.82582365077805819

In [7]:
cross_val_score(estimator = bernulli,X = breast_cancer.data,y = breast_cancer.target).mean()

0.62742040285899936

## Гаус

In [8]:
cross_val_score(estimator = gaussian,X = digit.data,y = digit.target).mean()

0.81860038035501381

In [9]:
cross_val_score(estimator = gaussian,X = breast_cancer.data,y = breast_cancer.target).mean()

0.9367492806089297

## Мультиномиальное

In [10]:
cross_val_score(estimator = multinomial,X = digit.data,y = digit.target).mean()

0.87087714897350532

In [11]:
cross_val_score(estimator = multinomial,X = breast_cancer.data,y = breast_cancer.target).mean()

0.89457904019307521

## Ответ

Для breast_cancer максимальным получилось 0.9367492806089297 Гаусовский наивный баес. Для digit - 0.87087714897350532 Мультиномиальный наивный баес. Верны утверждения (c) и (d).